### ISSUE TITLE
LanguageTree is not incremental

### ISSUE THREAD

**Issue by vigoux** [neovim/neovim#18107]
### Neovim version (nvim -v)

v0.7.0-dev+1436-gf0d07dcb7

### Vim (not Nvim) behaves the same?

Nope

### Operating system/version

Not relevant

### Terminal name/version

Not relevant

### $TERM environment variable

Not relevant

### Installation

Built from master

### How to reproduce the issue

Just use tree-sitter with a big file with lots of injections.
Editing the file will leads to catastrophic performances.

### Expected behavior

Better performance. This root of the problem is that the LanguageTree itself is not incremental, i.e. the childs of the language tree are recreated each time, and the queries are re-running each time too.

We could leverage the work in the highlighter and run the injection query on the updated ranges, possibly recreating children.

### Actual behavior

Not incremental.

**Comment by vigoux** [comment]
> The code base for language tree is also very complex with a lot of functionality integrated into. When people (including me) wanted to extend some functionality they most likely had to add code to multiple adding to already large `if` chain. Maybe small refactorings when tackling this problem would be good.
> 
> It appears that tree-sitter is handling the injections that there is only one parser for the whole document even for multiple injections spots that stays there and only injected regions are updated.

Let me rephrase how this is implemented today: there is only one language node per (parent language, language).

That is: if my language is `rust` and I have multiple `markdown` injections, there will only be one injected language node for all of them.

**Comment by keyvchan** [comment]
I would like to working on this, some instructions on where to get started would be very helpful. 😉

**Comment by vigoux** [comment]
Hey, please join the `#neovim-treesitter` channel on Matrix, we'll be going forward from here !

**Comment by MDeiml** [comment]
Some thoughts: I think there is no need to delete the old trees in
https://github.com/neovim/neovim/blob/f3ce06cfa139ca3fb142cf5adf96a2ecc4d8f551/runtime/lua/vim/treesitter/languagetree.lua#L289-L295
Tree sitter should be able to handle the changed included ranges.

Also the best thing would be to track invalidation per tree somehow and not for the whole language tree. That way only the trees that were edited could be invalidated in
https://github.com/neovim/neovim/blob/f3ce06cfa139ca3fb142cf5adf96a2ecc4d8f551/runtime/lua/vim/treesitter/languagetree.lua#L437
With these two changes on most edits only one or zero injected trees should be updated.

@keyvchan have you started working on this? Otherwise I'd be able to implement those changes?

**Comment by clason** [comment]
@MDeiml If you have a plan for it, go ahead!

**Comment by clason** [comment]
@MDeiml if you are still interested in working on this, that would be very much appreciated!

**Comment by MDeiml** [comment]
I haven't looked at this in quite some while, so I'm not that familiar anymore. But I believe the two changes I mentioned should be all that is needed. I'll see if I'll get to implementing that, but I can't promise anything.

**Comment by lewis6991** [comment]
#22309 should implement this. Would appreciate if anyone could test it.

For profiling I've added this to my init.lua:

```lua
local lt = require'vim.treesitter.languagetree'

local parse_time = 0
local nptrees = 0
local np = 0
local count = 0

local orig_parse_tree = lt.parse_tree -- added in #22309
lt.parse_tree = function(...)
  nptrees = nptrees+1
  return orig_parse_tree(...)
end

local orig_parse = lt.parse
lt.parse = function(...)
  local s = vim.loop.hrtime()
  np = np+1
  local r = {orig_parse(...)}
  parse_time = parse_time + vim.loop.hrtime() - s
  return unpack(r)
end

vim.api.nvim_create_autocmd('TextChanged', {
  callback = vim.schedule_wrap(function()
    count = count + 1
    print('P ', count, ' ', vim.inspect({
      parse = parse_time/1000000,
      parses_regions = nptrees,
      parses = np,
    }))
    parse_time = 0
    np = 0
    nptrees = 0
  end)
})

```

**Comment by clason** [comment]
Is this something that could go in a `make benchmark` test? Sounds too useful to keep hidden away in an issue comment...

**Comment by lewis6991** [comment]
We need a big file in our codebase that has lots of injections to test this with.

**Comment by clason** [comment]
or create one -- there's been too many "treesitter slow on my files" issues, so having a set of well-defined "stress test" files would be useful. (Doesn't need to be meaningful, just a, say, Lua file with a hundred pointless `vim.cmd` injections?)

**Comment by MDeiml** [comment]
Markdown uses injections for every inline content, so just a big markdown file should do. People have noted for example that [helix's changelog](https://raw.githubusercontent.com/helix-editor/helix/master/CHANGELOG.md) is slow to parse.

**Comment by clason** [comment]
The problem is that we don't bundle a Markdown parser currently (and adding one is not trivial since this would add a new dependency on a C++ compiler due to your external scanner). So it would have to be something using only Lua, Vimscript, Vimdoc (help), or C.
